packages/contracts-core/contracts/test/GovernanceMessage.t.sol

```diff
@@ -0,0 +1,488 @@
+    // array, not from the view
+    function test_identifierReadsFirstByte() public {
+        bytes memory data = hex"00";
+        assertEq(uint256(data.ref(0).identifier()), 0);
+        data = hex"01";
+        assertEq(uint256(data.ref(0).identifier()), 1);
+        data = hex"02";
+        assertEq(uint256(data.ref(0).identifier()), 2);
+    }
+
+    function test_serializeCall() public {
+        bytes32 to = "0xBEEF";
+        bytes memory data = "random data";
+        GovernanceMessage.Call memory call = GovernanceMessage.Call(to, data);
+        bytes29 serializedCall = abi
+            .encodePacked(to, uint32(data.length), data)
+            .ref(0);
+        assertEq(
+            GovernanceMessage.serializeCall(call).keccak(),
+            serializedCall.keccak()
+        );
+    }
+
+    function test_serializeCallFuzzed(bytes32 to, bytes memory data) public {
+        GovernanceMessage.Call memory call = GovernanceMessage.Call(to, data);
+        bytes29 serializedCall = abi
+            .encodePacked(to, uint32(data.length), data)
+            .ref(0);
+        assertEq(
+            GovernanceMessage.serializeCall(call).keccak(),
+            serializedCall.keccak()
+        );
+    }
+
+    function test_formatBatchSingleCall() public {
+        bytes32 to = "0xBEEF";
+        bytes memory data = "random data";
+        GovernanceMessage.Call[] memory calls = new GovernanceMessage.Call[](1);
+        calls[0] = GovernanceMessage.Call(to, data);
+        assertEq(
+            GovernanceMessage.formatBatch(calls),
+            abi.encodePacked(
+                GovernanceMessage.Types.Batch,
+                GovernanceMessage.getBatchHash(calls)
+            )
+        );
+    }
+
+    function test_formatBatchSingleCallFuzzed(bytes32 to, bytes memory data)
+        public
+    {
+        GovernanceMessage.Call[] memory calls = new GovernanceMessage.Call[](1);
+        calls[0] = GovernanceMessage.Call(to, data);
+        assertEq(
+            GovernanceMessage.formatBatch(calls),
+            abi.encodePacked(
+                GovernanceMessage.Types.Batch,
+                GovernanceMessage.getBatchHash(calls)
+            )
+        );
+    }
+
+    function test_getBatchHash() public {
+        // we only have 1 call
+        bytes memory prefix = hex"01";
+        bytes32 to = "0xBEEF";
+        bytes memory data = "random data";
+        bytes memory serializedCall = abi.encodePacked(
+            to,
+            uint32(data.length),
+            data
+        );
+        bytes memory batch = abi.encodePacked(prefix, serializedCall);
+        bytes32 batchHash = keccak256(batch);
+        GovernanceMessage.Call[] memory calls = new GovernanceMessage.Call[](1);
+        calls[0] = GovernanceMessage.Call(to, data);
+        assertEq(GovernanceMessage.getBatchHash(calls), batchHash);
+    }
+
+    function test_getBatchHashFuzzedIdenticalCalls(
+        bytes memory data,
```

Comment:
what are the fuzzer limitations on bytearrays? how varied will the length be? is it possible to do `bytes memory[]`